rain_api_core/auth.py

```diff
@@ -0,0 +1,154 @@
+            public_key: str,
+            private_key: str,
+            cookie_name: str,
+            blacklist_endpoint,
+            session_ttl_in_hours: int = 7 * 24
+    ):
+        self.algorithm = algorithm
+        self.public_key = public_key
+        self.private_key = private_key
+        self.cookie_name = cookie_name
+        self.black_list_endpoint = blacklist_endpoint
+        self.session_ttl = session_ttl_in_hours * 60 * 60
+        self.black_list_last_retrieved = 0
+        self.black_list = {}
+
+    def _get_auth_cookie(self, headers: Mapping[str, str]):
+        cookie_string = headers.get('cookie') or headers.get('Cookie') or headers.get('COOKIE')
+        if not cookie_string:
+            return {}
+
+        cookie = SimpleCookie()
+        with contextlib.suppress(CookieError):
+            cookie.load(cookie_string)
+        return cookie.get(self.cookie_name)
+
+    def _decode_jwt(self, token: str):
+        try:
+            return jwt.decode(token.encode(), self.public_key, self.algorithm)
+        except jwt.ExpiredSignatureError:
+            log.info('JWT has expired')
+        except jwt.InvalidSignatureError:
+            log.info('JWT has failed verification')
+        return None
+
+    def _encode_jwt(self, payload: Mapping[str, str]):
+        try:
+            encoded = jwt.encode(payload, self.private_key, self.algorithm)
+        except TypeError:
+            log.error('unable to encode jwt cookie')
+            return ''
+        return encoded
+
+    @staticmethod
+    def _format_date_for_cookie_header(timestamp):
+        return format_7231_date(timestamp)
+
+    def _jwt_payload_from_user_profile(self, user_profile: Union[UserProfile, None]):
+        if user_profile is None:
+            return {}
+        now = int(time())
+        return {
+            'urs-user-id': user_profile.user_id,
+            'first_name': user_profile.first_name,
+            'last_name': user_profile.last_name,
+            'email': user_profile.email,
+            'urs-access-token': user_profile.token,
+            'urs-groups': user_profile.groups,
+            'iat': now,
+            'exp': now + self.session_ttl
+        }
+
+    def _retrieve_blacklist(self):
+        if self.black_list and self.black_list_last_retrieved <= time() + (10 * 60):  # cached in last 10 minutes
+            pass
+        response = urllib.request.urlopen(self.black_list_endpoint).read().decode('utf-8')  # nosec
+        self.black_list = json.loads(response)['blacklist']
+        self.black_list_last_retrieved = time()
+
+    def _in_blacklist(self, user_profile: UserProfile):
+        self._retrieve_blacklist()
+        user_blacklist_time = self.black_list.get(user_profile.user_id)
+        if user_blacklist_time is not None:
+            if user_blacklist_time >= user_profile.iat:
+                return True
+        return False
+
+    def get_profile_from_headers(self, headers) -> Union[UserProfile, None]:
```

Comment:
Use `typing.Optional`
```suggestion
    def get_profile_from_headers(self, headers) -> Optional[UserProfile]:
```